vernac/vernacentries.ml

```diff
@@ -522,13 +523,17 @@ let start_lemma_com ~program_mode ~poly ~scope ~kind ?hook thms =
   in
   start_lemma_with_initialization ?hook ~poly ~scope ~kind evd ~udecl recguard thms snl
 
-let vernac_definition_hook ~local ~poly = let open Decls in function
+let vernac_definition_hook ~canonical_instance ~local ~poly = let open Decls in function
 | Coercion ->
   Some (Class.add_coercion_hook ~poly)
-| CanonicalStructure ->
-  Some (DeclareDef.Hook.(make (fun { S.dref } -> Canonical.declare_canonical_structure ?local dref)))
 | SubClass ->
   Some (Class.add_subclass_hook ~poly)
+| CanonicalStructure ->
+  Some (DeclareDef.Hook.(make (fun { S.dref } -> Canonical.declare_canonical_structure ?local dref)))
+| Definition when canonical_instance ->
+  Some (DeclareDef.Hook.(make (fun { S.dref } -> Canonical.declare_canonical_structure ?local dref)))
+| Let when canonical_instance ->
+  Some (DeclareDef.Hook.(make (fun { S.dref } -> Canonical.declare_canonical_structure dref)))
```

Comment:
3 identical branches, then please use an `or` pattern.

Reply:
You can't because 2 out of 3 have a when clause, and the 2 that have don't share the RHS

Reply:
Argh, indeed you cannot nest the when on the pattern.